article/_posts/2019-09-26-VQ-VAE.md

```diff
@@ -0,0 +1,163 @@
+error to move the make the latent code closer to the continuous vector it was
+matched to.
+
+Thus, the loss function has three components:
+
+![](/article/images/VQ-VAE/Loss_function.png)
+
+- The first term is the reconstruction loss (which optimizes the decoder and the
+encoder). Since a uniform prior for $$z$$ is assumed, the KL term that usually
+appears in the ELBO is constant w.r.t. the encoder parameters and can thus be
+ignored for training.
+- The second term is the contribution of the VQ, which optimises the embeddings.
+*sg* stands for "stop gradient" and is defined as identity at forward
+computation time and has zero partial derivatives, constraining its operand to
+be a non-updated constant.
+- The third term is a regularization term they name as *committment loss* which
+controls the volume of the latent space by forcing the encoder to *commit* to
+the latent code it matched with, and avoid its output space growing unboundedly.
+
+The second contribution of this work consists in learning the prior
+distribution. During the training phase, the prior $$p(z)$$ is a uniform
+categorical distribution. After the training is done, they fit an
+"autoregressive distribution" over the space of latent codes.
+
+
+# Results
+
+## Images
+
+Three complex image datasets (CIFAR10, ImageNet, and DeepMind Lab) were used.
+
+The proposed model is compared to the standard continuous VAE framework. It
+seems to achieve similar sample quality, while taking advantage of the discrete
+latent space. The model still performs well when using a PixelCNN decoder on top
+of the VQ-VAE.
+
+![](/article/images/VQ-VAE/Results_images.png)
+
+## Audio
+
+Experiments were done using the VCTK raw speech dataset and are available at:
+https://avdnoord.github.io/homepage/vqvae/
+
+They carried out experiments on the reconstruction of input clips from the
+low-dimensionality latent space reconstructions, as well as voice style transfer
+experiments, among others.
+
+Their model is able to factor speaker-specific information and to learn the same
+embeddings for the same (high-level) contents.
+
+## Video
+
+They used the DeepMind Lab environment to generate sequence of frames given a
+limited set of sequences. Their results show that the model can generate
+sequences without any degradation in the visual quality whilst keeping the local
+geometry correct.
+
+
+# Conclusions
+
+The VQ-VAE encoder network outputs discrete, rather than continuous, codes, and
+the prior is learned rather than being static.
+
+It is the first discrete latent VAE model that gets similar performance as its
+continuous counterparts.
+
+Advantages of their method:
+- It is simple to train.
+- It does not suffer from large variance.
+- It avoids the *posterior collapse* issue which has been problematic with many
+VAE models that have a powerful decoder, often caused by latents being ignored.
+
+
+# Comments
+
+The learning of the prior over the latent distribution is not clearly explained.
```

Comment:
From what I understand, the K discrete points form the domain of the prior distribution. However, it's not clear what probability would be assigned at those points.

Reply:
There may be some true in what we discussed during the reading group. Learning the latent space would mean learning the distribution, but we do not know which kind of distribution is learnt when the training ends; they state that they start with a uniform categorical distribution.